---
layout: post
title: 关于traceback研究
categories:
- python模块研究
tags:
- traceback
published: true
comments: true
indexer: true
---
<p>前言：traceback模块被用来跟踪异常返回信息</p>

<p>这事一个处理的脚本：</p>

<p><pre class="sh_python">
import sys</p>

<p>def produce_exception(recursion_level=2):
    sys.stdout.flush()
    if recursion_level:
        produce_exception(recursion_level-1)
    else:
        raise RuntimeError()</p>

<p>def call_function(f, recursion_level=2):
    if recursion_level:
        return call_function(f, recursion_level-1)
    else:
        return f()
</pre></p>

<p><pre class="sh_python">
import traceback
import sys
from pprint import pprint
from test1 import produce_exception  #引入上面脚本的方法</p>

<p>print 'print_exc() with no exception:'
traceback.print_exc(file=sys.stdout) #这里是none
print</p>

<p>try:
    produce_exception()
except Exception, err:
    print 'print_exc():'
    traceback.print_exc(file=sys.stdout)  #使用<tt>sys.exc_info()</tt> 获取当前线程,打印文本到一个文件句柄<tt>（默认是sys.stderr）</tt>
    print
    print 'print_exc(1):'
    traceback.print_exc(limit=1, file=sys.stdout)
    exc_type, exc_value, exc_tb = sys.exc_info()  
    traceback.print_exception(exc_type, exc_value, exc_tb) #其实traceback.print_exc是traceback.print_exception的短快捷方式 这2行的效果就是上面的
    pprint(traceback.format_exception(exc_type, exc_value, exc_tb)) #通过pprint以traceback.format_exception格式化的数据显示出来
</pre></p>

<p><pre class="sh_python">
import traceback
import sys
from pprint import pprint</p>

<p>from traceback_example import call_function</p>

<p>def f():
    traceback.print_stack(file=sys.stdout) #显示堆栈信息</p>

<p>print 'Calling f() directly:'
f()</p>

<p>print
print 'Calling f() from 3 levels deep:'
call_function(f)
def f():
    return traceback.format_stack() #将堆栈信息通过format_stack()格式化</p>

<p>formatted_stack = call_function(f)
pprint(formatted_stack)
def f(): 
    return traceback.extract_stack()  #提取堆栈信息
stack = call_function(f)
pprint(stack)
</pre></p>
